pay/contracts/refund/HectorRefund.sol

```diff
@@ -0,0 +1,128 @@
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize() external initializer {
+        moderators[msg.sender] = true;
+
+        __Ownable_init();
+    }
+
+    /* ======== MODIFIER ======== */
+
+    modifier onlyMod() {
+        if (!moderators[msg.sender]) revert INVALID_MODERATOR();
+        _;
+    }
+
+    /* ======== POLICY FUNCTIONS ======== */
+
+    function setModerator(address moderator, bool approved) external onlyOwner {
+        if (moderator == address(0)) revert INVALID_ADDRESS();
+        moderators[moderator] = approved;
+    }
+
+    /* ======== MODERATOR FUNCTIONS ======== */
+
+    function appendRefund(
+        uint256[] calldata _planIds,
+        Refund[][] calldata _refunds
+    ) external onlyMod {
+        uint256 length = _planIds.length;
+        if (length != _refunds.length) revert INVALID_PARAM();
+
+        for (uint256 i = 0; i < length; i++) {
+            uint256 planId = _planIds[i];
+            uint256 subLength = _refunds[i].length;
+
+            for (uint256 j = 0; j < subLength; j++) {
+                Refund memory refund = _refunds[i][j];
+                if (refund.limitPeriod == 0) revert INVALID_PERIOD();
+                if (refund.percent > MULTIPLIER) revert INVALID_PERCENT();
+
+                refunds[planId].push(refund);
+            }
+        }
+    }
+
+    function updateRefund(
+        uint256 _planId,
+        uint256 _index,
+        Refund calldata _refund
+    ) external onlyMod {
+        if (_index >= refunds[_planId].length) revert INVALID_PARAM();
+        if (_refund.limitPeriod == 0) revert INVALID_PERIOD();
+        if (_refund.percent > MULTIPLIER) revert INVALID_PERCENT();
+
+        refunds[_planId][_index] = _refund;
+    }
+
+    function removeRefund(uint256 _planId, uint256 _index) external onlyMod {
+        uint256 length = refunds[_planId].length;
+        if (_index >= length) revert INVALID_PARAM();
+
+        refunds[_planId][_index] = refunds[_planId][length - 1];
+        delete refunds[_planId][length - 1];
+        refunds[_planId].pop();
+    }
+
+    /* ======== VIEW FUNCTIONS ======== */
+
+    function allRefundsForPlan(
+        uint256 _planId
+    ) external view returns (Refund[] memory) {
+        return refunds[_planId];
+    }
+
+    function applyRefund(
+        bytes calldata subscription
+    ) external view returns (uint256 refundAmount) {
+        (uint256 planId, uint48 lastPaidAt, uint256 lastPaidAmount) = abi
+            .decode(subscription, (uint256, uint48, uint256));
+
+        uint48 period = uint48(block.timestamp) - lastPaidAt;
+        uint48 limitPeriod = 2 ** 48 - 1;
+        uint48 percent = 0;
+        uint256 length = refunds[planId].length;
+
+        for (uint256 i = 0; i < length; i++) {
+            Refund memory refund = refunds[planId][i];
+
+            if (
+                limitPeriod < refund.limitPeriod && period < refund.limitPeriod
```

Comment:
This refund data will be prepared for each plan

Reply:
let me explain how the refund contract is handling our cases
$150 for 3months
- On or before 15 days, 150 available for refund
Refund {limitPeriod: 15 days, percent: 100%}

- After 15 days, 100 available for refund
Refund (limitPeriod: 30 days, percent: 66.66%)

- After 31 days, 50 available for refund
Refund (limitPeriod: 60 days, percent: 33.33%)

- After 61 days, 0 available for refund
0%

Contract mechanism: getRefundAmount in 14 days
find (14 days < refund.limitPeriod) and max percent
So it would be 100% and all paid will be refunded